js/modules/loki_app_dot_net_api.js

```diff
@@ -800,104 +838,222 @@ class LokiPublicChannelAPI {
-              id: this.conversationId,
-              type: textsecure.protobuf.GroupContext.Type.DELIVER,
+          // Filter out any messages that we got previously
+          if (this.lastMessagesCache.some(isDuplicate)) {
+            return false; // Duplicate message
+          }
+
+          // FIXME: maybe move after the de-multidev-decode
+          // Add the message to the lastMessage cache and keep the last 5 recent messages
+          this.lastMessagesCache = [
+            ...this.lastMessagesCache,
+            {
+              username: adnMessage.user.username,
+              text: adnMessage.text,
+              timestamp,
             },
-            flags: 0,
-            expireTimer: 0,
-            profileKey: null,
+          ].splice(-5);
+
+          const from = adnMessage.user.name || 'Anonymous'; // profileName
+          if (pubKeys.indexOf(`@${adnMessage.user.username}`) === -1) {
+            pubKeys.push(`@${adnMessage.user.username}`);
+          }
+
+          const messageData = {
+            serverId: adnMessage.id,
+            clientVerified: true,
+            friendRequest: false,
+            source: adnMessage.user.username,
+            sourceDevice: 1,
             timestamp,
-            received_at: receivedAt,
-            sent_at: timestamp,
-            quote,
-            contact: [],
-            preview,
-            profile: {
-              displayName: from,
-            },
-          },
-        };
-        receivedAt += 1; // Ensure different arrival times
 
-        this.serverAPI.chatAPI.emit('publicMessage', {
-          message: messageData,
-        });
+            serverTimestamp: timestamp,
+            receivedAt,
+            isPublic: true,
+            message: {
+              body:
+                adnMessage.text === timestamp.toString() ? '' : adnMessage.text,
+              attachments,
+              group: {
+                id: this.conversationId,
+                type: textsecure.protobuf.GroupContext.Type.DELIVER,
+              },
+              flags: 0,
+              expireTimer: 0,
+              profileKey: null,
+              timestamp,
+              received_at: receivedAt,
+              sent_at: timestamp,
+              quote,
+              contact: [],
+              preview,
+              profile: {
+                displayName: from,
+              },
+            },
+          };
+          receivedAt += 1; // Ensure different arrival times
 
-        // now process any user meta data updates
-        // - update their conversation with a potentially new avatar
-      });
+          // now process any user meta data updates
+          // - update their conversation with a potentially new avatar
+          return messageData;
+        })
+      );
       this.conversation.setLastRetrievedMessage(this.lastGot);
+
+      if (pendingMessages.length) {
+        // console.log('premultiDeviceResults', pubKeys);
+        if (pubKeys.length) {
```

Comment:
If you just feed pubKeys in to
```js
const verifiedPrimaryPKs = await lokiFileServerAPI.verifyPrimaryPubKeys(
  pubKeys
);
```
after pendingMessages has been generated above (pubKeys will always be at least an empty array and verifyPrimaryPubKeys can handle that) we can flatten some more here by at least just replacing this check with
```js
if (verifiedPrimaryPKs.length) {
```
and removing that check from below